src/main/kotlin/net/maizegenetics/phgv2/cli/CreateRanges.kt

```diff
@@ -1,16 +1,115 @@
 
-    override fun run() {
-        // TBD
+    /**
+     * Identifies minimum and maximum positions for a list of genes
+     *
+     * @param genes A list of type `Gene`
+     * @param boundary A type of boundary. Either `gene` or `cds`
+     * @param pad Number of base-pairs to flank boundary regions
+     *
+     * @return A list of type `Pair<Int, Int>`
+     */
+    fun idMinMaxBounds(genes: List<Gene>, boundary: String, pad: Int): List<Pair<Int, Int>> {
+        val boundMinMax = when (boundary) {
+            "gene" -> genes.map {gene ->
+                Pair(gene.start, gene.end)
+            }
+            "cds" -> {
+                genes.map { gene ->
+                    gene.children
+                        .filter { transcript ->
+                            transcript.attributes.containsKey("canonical_transcript")
+                        }
+                        .flatMap { it.children }
+                        .filter { it.type().name == "CODING_SEQUENCE" }
+                        .map { Pair(it.start, it.end) }
+                        .let { ranges ->
+                            Pair(ranges.minOf { it.first }, ranges.maxOf { it.second })
+                        }
+                }
+            }
+            else -> throw Exception("Undefined boundary")
+        }.map { (first, second) ->
+            val modifiedFirst = if (first - pad < 0) 1 else first - pad
+            Pair((modifiedFirst - 1), (second + pad) - 1)
+        }
+
+        return(boundMinMax)
+    }
+
+    /**
+     * Generates a list of BED-formatted string rows
+     *
+     * @param bounds A list of type `Pair<Int, Int>`. Generated from [idMinMaxBounds]`.
+     * @param genes A list of type `Gene`
+     * @param delimiter A column delimiter for a BED file. Defaults to `\t`
+     * @param featureId Identifier key for value to pull from ID field of GFF. Defaults to `"ID"`
+     *
+     * @return A list of type `String`
+     */
+    fun generateBedRows(
+        bounds: List<Pair<Int, Int>>,
+        genes: List<Gene>,
+        delimiter: String = "\t",
+        featureId: String = "ID"
+    ): List<String> {
+        val bedLinesToPrint = genes.mapIndexed { index, gene ->
+            listOf(
+                gene.seqid,
+                bounds[index].first,
+                bounds[index].second,
+                gene.attributes[featureId],
+                if (gene.score.isNaN()) 0 else gene.score,
+                gene.strand
+            ).joinToString(delimiter)
+        }
+
+        return(bedLinesToPrint)
     }
 
-}
\ No newline at end of file
+    override fun run() {
+        val genome = Genome.fromGFF(gff)
+        val genes = genome.genes()
+
+        val boundMinMax = idMinMaxBounds(genes, boundary, pad)
+
+        val bedLinesToPrint = generateBedRows(boundMinMax, genes)
+
+        val bw = output?.let { File(it).bufferedWriter() }
+        bedLinesToPrint.forEach {
+            bw?.write("$it\n")
+        }
+        bw?.close()
```

Comment:
Might be better to do use {} for automatic closing

Reply:
The output file name here is optional and I think the let makes it a bit easier.  I changed it to use .use{} but now my code coverage is lower so I am working through adding a unit test.